### PR TITLE
Blueprint zones for freeform construction planning.

### PR DESCRIPTION
**Pull request by ghost**
#### Summary
```SUMMARY: Features "Blueprint zones for freeform construction planning"```

#### Purpose of change
Allow blueprint zones to specify areas of construction for different types of construction for NPCs to autonomously work on.

#### Describe the solution
A new zone type - Blueprint , where you specify a construction type in that zone.
Then you ask your NPC  to go work on blueprints.
If the componnts and tools are near to the zone, they will construct it, and then move on to the next one in the zone.
They will also build the pre-requisite if needed, for example, a fence requires fence posts first, if you specify a fence zone, they will build the fence posts first, then the fence.

#### Describe alternatives you've considered
I eventually want them to link this into basecamp jobs/missions, so when not on a mission they do idle tasks like this and crafting, and sorting, and butchering, thatll come soon.

I didn't mnage to work out how best to make them go and fetch the components and take them to the build site, for the moment they rely on you, ( or NPC sorting ) for the components to be within crafting range of the build site.

#### Additional context
The performance will probably be quite poor with this,
they check if they can build the thing before moving there, then again when they arrive to see if anythings changed in the meantime.
Im sure there are more performant and more elegant ways to do this, but as usual, im putting the basic implementation up first, and then we can work on prettying it up.

I've allowed the player to use the zone too, with shift+o in a blueprint zone, theyll attempt to to the same thing.

### CODE REVIEW

File: data/json/npcs/TALK_COMMON_ALLY.json [hunk 1]
```diff
@@ -722,10 +722,16 @@
         "effect": "sort_loot"
       },
       {
-        "text": "Please work on any construction taks that you know how to finish.",
+        "text": "Please work on any unfinished construction taks that you know how to finish.",
```
Comment (Night-Pryanik) [inline data/json/npcs/TALK_COMMON_ALLY.json]:
```suggestion
        "text": "Please work on any unfinished construction task that you know how to finish.",
```

File: src/activity_item_handling.cpp [hunk 2]
```diff
@@ -1017,6 +1019,249 @@ std::vector<tripoint> route_adjacent( const player &p, const tripoint &dest )
...
+                adjacent = true;
+                break;
+            }
+        }
+        if( !adjacent ) {
+            std::vector<tripoint> route = route_adjacent( p, src_loc );
+
+            // check if we found path to source / adjacent tile
+            if( route.empty() ) {
+                add_msg( m_info, _( "%s can't reach the source tile to construct" ),
```
Comment (Night-Pryanik) [inline src/activity_item_handling.cpp]:
```suggestion
                add_msg( m_info, _( "%s can't reach the source tile to construct." ),
```

**Comment by I-am-Erk** [comment]
Ah. This could be how we do npc repairs of facilities actually. If I can place these zones in mapgen I can define what furnishings NPCs should go around keeping fixed up to blueprint.

**Comment by davidpwbrown** [comment]
> Ah. This could be how we do npc repairs of facilities actually. If I can place these zones in mapgen I can define what furnishings NPCs should go around keeping fixed up to blueprint.

Yep, I could work out how to place them in mapgen like ownership zones in some way. Thatd be neat.

**Comment by Metalsoul212** [comment]
To expand on Erks idea. I wonder if this could be use by worldgen NPCs in conjunction with Kevin's proposed AI tree. When the NPC seeks shelter they could maybe use this mapgen blueprint zone idea as a way to fortify and expand their shelter, based on the building they chose!

**Comment by davidpwbrown** [comment]
> To expand on Erks idea. I wonder if this could be use by worldgen NPCs in conjunction with Kevin's proposed AI tree. When the NPC seeks shelter they could maybe use this mapgen blueprint zone idea as a way to fortify and expand their shelter, based on the building they chose!

Thats the plan yes, to make factions self-sufficient and build up organically.

**Comment by mlangsdorf** [comment]
I think we should add a boolean `"set_blueprint_zone"` to mapgen definitions.  If true, after mapgen it would set the construction blueprint zones automatically.  Then map creators wouldn't have to do any additional work.

There would still be an option to set up construction zones manually if you only wanted part of the map tile to have zones, but this would simplify things for large zones.

dpwb, I'll review this more tomorrow.  There's probably some overlap between the modular faction camp code and this.

**Comment by I-am-Erk** [comment]
That's a good idea, it might be nice if it could be specified both for terrain and for furniture though. Specifically thinking of the piles of furniture around the back bay in the refugee center, I don't really want something like that to get constantly replaced.

That said, I'm going to lock the doors and remove the barriers pretty soon here, and unlock the doors when you get the quest to go there, so that specific example of the problem isn't a big deal.

**Comment by mlangsdorf** [comment]
Fixes #26211

**Comment by davidpwbrown** [comment]
oh ffs messed something up

File: src/activity_item_handling.cpp [hunk 3]
```diff
@@ -1017,6 +1019,256 @@ std::vector<tripoint> route_adjacent( const player &p, const tripoint &dest )
     return std::vector<tripoint>();
 }
 
+static construction check_build_pre( const construction &con )
+{
+    std::string pre_con_str = con.pre_terrain;
```
Comment (KorGgenT) [inline src/activity_item_handling.cpp]:
```suggestion
    const std::string pre_con_str = con.pre_terrain;
```

File: src/activity_item_handling.cpp [hunk 4]
```diff
@@ -1017,6 +1019,256 @@ std::vector<tripoint> route_adjacent( const player &p, const tripoint &dest )
...
+            pre_con = elem;
+            break;
+        }
+    }
+    return pre_con;
+}
+
+void activity_on_turn_blueprint_move( player_activity &, player &p )
+{
+    auto &mgr = zone_manager::get_manager();
```
Comment (KorGgenT) [inline src/activity_item_handling.cpp]:
```suggestion
    zone_manager &mgr = zone_manager::get_manager();
```

File: src/activity_item_handling.cpp [hunk 5]
```diff
@@ -1017,6 +1019,256 @@ std::vector<tripoint> route_adjacent( const player &p, const tripoint &dest )
...
+
+    const auto abspos = g->m.getabs( p.pos() );
```
Comment (KorGgenT) [inline src/activity_item_handling.cpp]:
```suggestion
    const tripoint abspos = g->m.getabs( p.pos() );
```

File: src/activity_item_handling.cpp [hunk 6]
```diff
@@ -1017,6 +1019,256 @@ std::vector<tripoint> route_adjacent( const player &p, const tripoint &dest )
...
+    const auto &src_set = mgr.get_near( zone_type_id( "CONSTRUCTION_BLUEPRINT" ), abspos );
```
Comment (KorGgenT) [inline src/activity_item_handling.cpp]:
```suggestion
    const std::unordered_set<tripoint> &src_set = mgr.get_near( zone_type_id( "CONSTRUCTION_BLUEPRINT" ), abspos );
```

File: src/activity_item_handling.cpp [hunk 7]
```diff
@@ -1017,6 +1019,256 @@ std::vector<tripoint> route_adjacent( const player &p, const tripoint &dest )
...
+
+    const auto &src_sorted = get_sorted_tiles_by_distance( abspos, src_set );
```
Comment (KorGgenT) [inline src/activity_item_handling.cpp]:
```suggestion
    const std::vector<tripoint> &src_sorted = get_sorted_tiles_by_distance( abspos, src_set );
```

File: src/activity_item_handling.cpp [hunk 8]
```diff
@@ -1017,6 +1019,256 @@ std::vector<tripoint> route_adjacent( const player &p, const tripoint &dest )
...
+    const activity_id act_multiple_construction = activity_id( "ACT_BLUEPRINT_CONSTRUCTION" );
+
+    // Nuke the current activity, leaving the backlog alone
+    p.activity = player_activity();
+
+    // sort source tiles by distance
+    for( auto &src : src_sorted ) {
```
Comment (KorGgenT) [inline src/activity_item_handling.cpp]:
```suggestion
    for( const tripoint &src : src_sorted ) {
```

File: src/activity_item_handling.cpp [hunk 9]
```diff
@@ -1017,6 +1019,256 @@ std::vector<tripoint> route_adjacent( const player &p, const tripoint &dest )
...
+        const auto &src_loc = g->m.getlocal( src );
```
Comment (KorGgenT) [inline src/activity_item_handling.cpp]:
```suggestion
        const tripoint &src_loc = g->m.getlocal( src );
```

File: src/activity_item_handling.cpp [hunk 10]
```diff
@@ -1017,6 +1019,256 @@ std::vector<tripoint> route_adjacent( const player &p, const tripoint &dest )
...
+        if( !g->m.inbounds( src_loc ) ) {
+            if( !g->m.inbounds( p.pos() ) ) {
+                // p is implicitly an NPC that has been moved off the map, so reset the activity
+                // and unload them
+                p.assign_activity( act_multiple_construction );
+                p.set_moves( 0 );
+                g->reload_npcs();
+                return;
+            }
+            std::vector<tripoint> route = route_adjacent( p, src_loc );
```
Comment (KorGgenT) [inline src/activity_item_handling.cpp]:
```suggestion
            const std::vector<tripoint> route = route_adjacent( p, src_loc );
```

File: src/activity_item_handling.cpp [hunk 11]
```diff
@@ -1017,6 +1019,256 @@ std::vector<tripoint> route_adjacent( const player &p, const tripoint &dest )
...
+            if( route.empty() ) {
+                // can't get there, can't do anything, skip it
+                continue;
+            }
+            p.set_destination( route, player_activity( act_multiple_construction ) );
+            return;
+        }
+        // dont go there if it's dangerous.
+        bool dangerous_field = false;
+        for( const auto &e : g->m.field_at( src_loc ) ) {
```
Comment (KorGgenT) [inline src/activity_item_handling.cpp]:
```suggestion
        for( const std::pair<const field_id, field_entry> &e : g->m.field_at( src_loc ) ) {
```

File: src/activity_item_handling.cpp [hunk 12]
```diff
@@ -1017,6 +1019,256 @@ std::vector<tripoint> route_adjacent( const player &p, const tripoint &dest )
...
+            if( p.is_dangerous_field( e.second ) ) {
+                dangerous_field = true;
+                break;
+            }
+        }
+        if( dangerous_field ) {
+            continue;
+        }
+        // work out if we can build it before we move there.
+        const auto &zones = mgr.get_zones( zone_type_id( "CONSTRUCTION_BLUEPRINT" ),
```
Comment (KorGgenT) [inline src/activity_item_handling.cpp]:
```suggestion
        const std::vector<zone_data> &zones = mgr.get_zones( zone_type_id( "CONSTRUCTION_BLUEPRINT" ),
```

File: src/activity_item_handling.cpp [hunk 13]
```diff
@@ -1017,6 +1019,256 @@ std::vector<tripoint> route_adjacent( const player &p, const tripoint &dest )
...
+                                           g->m.getabs( src_loc ) );
+        construction built_chosen;
+        const inventory pre_inv = p.crafting_inventory( src_loc, PICKUP_RANGE - 1 );
+        // PICKUP_RANGE -1 because we will be adjacent to the spot when arriving.
+        bool found_any_pre = false;
+
+        for( const auto &zone : zones ) {
```
Comment (KorGgenT) [inline src/activity_item_handling.cpp]:
```suggestion
        for( const zone_data &zone : zones ) {
```

File: src/activity_item_handling.cpp [hunk 14]
```diff
@@ -1017,6 +1019,256 @@ std::vector<tripoint> route_adjacent( const player &p, const tripoint &dest )
...
+            const auto options = dynamic_cast<const blueprint_options &>( zone.get_options() );
```
Comment (KorGgenT) [inline src/activity_item_handling.cpp]:
```suggestion
            const blueprint_options options = dynamic_cast<const blueprint_options &>( zone.get_options() );
```

File: src/activity_item_handling.cpp [hunk 15]
```diff
@@ -1017,6 +1019,256 @@ std::vector<tripoint> route_adjacent( const player &p, const tripoint &dest )
...
+                    }
+                }
+                continue;
+            }
+        }
+        if( !found_any_pre ) {
+            continue;
+        }
+        bool adjacent = false;
+        for( auto elem : g->m.points_in_radius( src_loc, 1 ) ) {
```
Comment (KorGgenT) [inline src/activity_item_handling.cpp]:
```suggestion
        for( const tripoint &elem : g->m.points_in_radius( src_loc, 1 ) ) {
```

File: src/activity_item_handling.cpp [hunk 16]
```diff
@@ -1017,6 +1019,256 @@ std::vector<tripoint> route_adjacent( const player &p, const tripoint &dest )
...
+            p.set_destination( route, player_activity( act_multiple_construction ) );
+            return;
+        }
+        // if it's too dark to construct there
+        const bool enough_light = p.fine_detail_vision_mod() <= 4;
+        if( !enough_light ) {
+            continue;
+        }
+        // check if can do the construction now we are actually there
+        const auto &post_zones = mgr.get_zones( zone_type_id( "CONSTRUCTION_BLUEPRINT" ),
```
Comment (KorGgenT) [inline src/activity_item_handling.cpp]:
```suggestion
        const std::vector<zone_data> &post_zones = mgr.get_zones( zone_type_id( "CONSTRUCTION_BLUEPRINT" ),
```

File: src/activity_item_handling.cpp [hunk 17]
```diff
@@ -1017,6 +1019,256 @@ std::vector<tripoint> route_adjacent( const player &p, const tripoint &dest )
...
+                                                g->m.getabs( src_loc ) );
+        construction post_built_chosen;
+        p.invalidate_crafting_inventory();
+        const inventory &total_inv = p.crafting_inventory();
+        bool found_any = false;
+
+        for( const auto &zone : post_zones ) {
```
Comment (KorGgenT) [inline src/activity_item_handling.cpp]:
```suggestion
        for( const zone_data &zone : post_zones ) {
```

File: src/activity_item_handling.cpp [hunk 18]
```diff
@@ -1017,6 +1019,256 @@ std::vector<tripoint> route_adjacent( const player &p, const tripoint &dest )
...
+            const auto options = dynamic_cast<const blueprint_options &>( zone.get_options() );
```
Comment (KorGgenT) [inline src/activity_item_handling.cpp]:
```suggestion
            const blueprint_options options = dynamic_cast<const blueprint_options &>( zone.get_options() );
```

File: src/activity_item_handling.cpp [hunk 19]
```diff
@@ -1017,6 +1019,256 @@ std::vector<tripoint> route_adjacent( const player &p, const tripoint &dest )
...
+        }
+        std::list<item> used;
+        // create the partial construction struct
+        partial_con pc;
+        pc.id = built_chosen.id;
+        pc.counter = 0;
+        // Set the trap that has the examine function
+        g->m.trap_set( src_loc, tr_unfinished_construction );
+        // Use up the components
+        for( const auto &it : built_chosen.requirements->get_components() ) {
```
Comment (KorGgenT) [inline src/activity_item_handling.cpp]:
```suggestion
        for( const std::vector<item_comp> &it : built_chosen.requirements->get_components() ) {
```

File: src/activity_item_handling.cpp [hunk 20]
```diff
@@ -1017,6 +1019,256 @@ std::vector<tripoint> route_adjacent( const player &p, const tripoint &dest )
...
+            std::list<item> tmp = p.consume_items( it, 1, is_crafting_component );
+            used.splice( used.end(), tmp );
+        }
+        pc.components = used;
+        g->m.partial_con_set( src_loc, pc );
+        for( const auto &it : built_chosen.requirements->get_tools() ) {
```
Comment (KorGgenT) [inline src/activity_item_handling.cpp]:
```suggestion
        for( const std::vector<tool_comp> &it : built_chosen.requirements->get_tools() ) {
```

File: src/clzones.cpp [hunk 21]
```diff
@@ -236,6 +289,15 @@ std::string plot_options::get_zone_name_suggestion() const
     return _( "No seed" );
 }
 
+std::vector<std::pair<std::string, std::string>> blueprint_options::get_descriptions() const
+{
+    auto options = std::vector<std::pair<std::string, std::string>>();
```
Comment (KorGgenT) [inline src/clzones.cpp]:
```suggestion
    std::vector<std::pair<std::string, std::string>> options = std::vector<std::pair<std::string, std::string>>();
```

File: src/activity_item_handling.cpp [hunk 22]
```diff
@@ -1017,6 +1019,256 @@ std::vector<tripoint> route_adjacent( const player &p, const tripoint &dest )
     return std::vector<tripoint>();
 }
 
+static construction check_build_pre( const construction &con )
+{
+    std::string pre_con_str = con.pre_terrain;
+    construction pre_con = con;
+    const std::vector<construction> &list_constructions = get_constructions();
+    for( const auto elem : list_constructions ) {
```
Comment (KorGgenT) [inline src/activity_item_handling.cpp]:
```suggestion
    for( const construction elem : list_constructions ) {
```

**Comment by davidpwbrown** [comment]
auto is ded.

File: src/activity_item_handling.cpp [hunk 23]
```diff
@@ -1017,6 +1019,256 @@ std::vector<tripoint> route_adjacent( const player &p, const tripoint &dest )
...
+    }
+    return pre_con;
+}
+
+void activity_on_turn_blueprint_move( player_activity &, player &p )
+{
+    zone_manager &mgr = zone_manager::get_manager();
+
+    const tripoint abspos = g->m.getabs( p.pos() );
+    const std::unordered_set<tripoint> &src_set = mgr.get_near( zone_type_id( "CONSTRUCTION_BLUEPRINT" ), abspos );
```
Comment (ZhilkinSerg) [inline src/activity_item_handling.cpp]:
```suggestion
    const std::unordered_set<tripoint> &src_set = mgr.get_near(
                zone_type_id( "CONSTRUCTION_BLUEPRINT" ), abspos );
```

File: src/activity_item_handling.cpp [hunk 24]
```diff
@@ -1017,6 +1019,256 @@ std::vector<tripoint> route_adjacent( const player &p, const tripoint &dest )
...
+                dangerous_field = true;
+                break;
+            }
+        }
+        if( dangerous_field ) {
+            continue;
+        }
+        // work out if we can build it before we move there.
+        const std::vector<zone_data> &zones = mgr.get_zones( zone_type_id( "CONSTRUCTION_BLUEPRINT" ),
+                                           g->m.getabs( src_loc ) );
```
Comment (ZhilkinSerg) [inline src/activity_item_handling.cpp]:
```suggestion
                                              g->m.getabs( src_loc ) );
```

File: src/activity_item_handling.cpp [hunk 25]
```diff
@@ -1017,6 +1019,256 @@ std::vector<tripoint> route_adjacent( const player &p, const tripoint &dest )
...
+            return;
+        }
+        // if it's too dark to construct there
+        const bool enough_light = p.fine_detail_vision_mod() <= 4;
+        if( !enough_light ) {
+            continue;
+        }
+        // check if can do the construction now we are actually there
+        const std::vector<zone_data> &post_zones = mgr.get_zones( zone_type_id( "CONSTRUCTION_BLUEPRINT" ),
+                                                g->m.getabs( src_loc ) );
```
Comment (ZhilkinSerg) [inline src/activity_item_handling.cpp]:
```suggestion
                g->m.getabs( src_loc ) );
```

File: src/clzones.cpp [hunk 26]
```diff
@@ -236,6 +289,15 @@ std::string plot_options::get_zone_name_suggestion() const
     return _( "No seed" );
 }
 
+std::vector<std::pair<std::string, std::string>> blueprint_options::get_descriptions() const
+{
+    std::vector<std::pair<std::string, std::string>> options = std::vector<std::pair<std::string, std::string>>();
```
Comment (ZhilkinSerg) [inline src/clzones.cpp]:
```suggestion
    std::vector<std::pair<std::string, std::string>> options =
                std::vector<std::pair<std::string, std::string>>();
```

Comment (ifreund) [reply src/clzones.cpp]: there's no need to call the constructor like that.  Should just be
```cpp
std::vector<std::pair<std::string, std::string>> options;
```

**Pull request merged**